### ISSUE TITLE
Current status of lifetime management?

### ISSUE THREAD

**Issue by Ziriax** [sharpdx/SharpDX#967]
In the early days of SharpDX, it was adviced to explicitly call `Dispose`, not allowing the garbage collector to perform cleanup in the finalizers. Since the .NET runtime doesn't have a concept of GPU memory pressure, I assume this is still the case, but in the past, this often just crashed. This is even stated in the [DirectX FAQ](https://msdn.microsoft.com/en-us/library/windows/desktop/ee416788(v=vs.85).aspx#DOES_IT_MATTER_IN_WHICH_ORDER_I_RELEASE_DIRECTX_INTERFACES__): 

> **Does it matter in which order I release DirectX interfaces?** It shouldn't matter because COM interfaces are reference counted. _**However, there are some known bugs with the release order of interfaces in some versions of DirectX**_. For safety, you are advised to release interfaces in reverse creation order when possible.

What is current status on this? Since we are building a playground for experimenting, we don't care that much about device resources immediately being released, and we would like to just use the garbage collector.

Furthermore, often SharpDX creates new COM objects by just using a property, making memory leaks very easy to make.

So, it seems that DirectX 11 still requires some order in releasing the objects. At least that is what I'm concluding from my experiments, but I could be wrong. For example, it seems one must first release device-children before releasing the device (although each device-child has a reference to the device, this is a weak COM reference, so the child does not increment the ref-count of the device). But finalizers do not guarantee any order of execution...

A problem with the manual `Dispose` approach is that one must always be able to assign a unique owner .NET class that will call `Dispose`. This is not always possible. In COM, we just use reference counts for this. I tried a similar approach in C#, calling `IUnknown.AddReference`  and `IUnknown.Release`, and then when the reference count becomes zero, setting the `ComObject.NativePointer` to zero, and calling `ComObject.Dispose`. This works most of the time, but I have some SharpDX crashes because not all code expects `NativePointer` to be zero.

In C++ this is all very easy, just use smart pointers, but just about everything else is a lot easier in C# ;-)

So to summarize, what is the correct way of dealing with shared resources in SharpDX, and can one use garbage collector finalizers these days for automatic native resource lifetime management?

Thanks a lot,
Peter Verswyvelen

**Comment by xoofx** [comment]
> What is current status on this?

The advice still prevails. Calling Dispose on SharpDX object should always work. And indeed, the feature of auto-disposing has been made optional in SharpDX (you can still activate it via [Configuration](https://github.com/sharpdx/SharpDX/blob/master/Source/SharpDX/Configuration.cs)), but as stated I don't guarantee that it will always work.

Also, as you said, the Dispose pattern is mostly fine for single ownership, so if you can stick with this pattern it is better. If you can't, you have either to track things with AddRef/ReleaseRef (which is cumbersome) or let auto-dispose working.

**Comment by Ziriax** [comment]
Thanks! We really need destructible value types in C# so we can have smart pointers :-(
